### PR TITLE
duplicate node

### PR DESCRIPTION
**Pull request by AbdelHedhili**
Signed-off-by: Abdelsalem <abdelsalem.hedhili@rte-france.com>

### CODE REVIEW

File: src/components/graph/menus/create-node-menu.js [hunk 1]
```diff
@@ -90,6 +103,29 @@ const CreateNodeMenu = ({
...
+        },
+        PASTE_MODIFICATION_NODE_AFTER: {
+            onRoot: true,
+            action: () => pasteNetworkModificationNode('AFTER'),
+            id: 'pasteNetworkModificationNodeBelow',
+            disabled: selectedNodeForCopy == null,
+        },
+        PASTE_MODIFICATION_NODE_ON_NEW_BRANCH: {
+            onRoot: false,
+            action: () => pasteNetworkModificationNode('NEW_BRANCH'),
```
Comment (FranckLecuyer) [inline src/components/graph/menus/create-node-menu.js:125]: Same remark in study-server PR : 
CHILD and NEW_BRANCH seems to have same signification ...
maybe remove CHILD ?

**Review by FranckLecuyer — CHANGES_REQUESTED** [review]

**Review by SlimaneAmar — DISMISSED** [review]

File: src/components/graph/menus/create-node-menu.js [hunk 2]
```diff
@@ -90,6 +103,29 @@ const CreateNodeMenu = ({
...
+            id: 'pasteNetworkModificationNodeAbove',
+            disabled: selectedNodeForCopy == null,
+        },
+        PASTE_MODIFICATION_NODE_AFTER: {
+            onRoot: true,
+            action: () => pasteNetworkModificationNode('AFTER'),
+            id: 'pasteNetworkModificationNodeBelow',
+            disabled: selectedNodeForCopy == null,
+        },
+        PASTE_MODIFICATION_NODE_ON_NEW_BRANCH: {
```
[the remark this answers is not in the document]
Comment (AbdelHedhili) [reply src/components/graph/menus/create-node-menu.js:123]: done

File: src/components/graph/menus/create-node-menu.js [hunk 3]
```diff
@@ -90,6 +103,29 @@ const CreateNodeMenu = ({
             action: () => createNetworkModificationNode('AFTER'),
             id: 'insertNetworkModificationNodeBelow',
         },
+        COPY_MODIFICATION: {
```
[the remark this answers is not in the document]
Comment (AbdelHedhili) [reply src/components/graph/menus/create-node-menu.js:106]: done

File: src/components/network-modification-tree-pane.js [hunk 4]
```diff
@@ -80,6 +81,7 @@ export const NetworkModificationTreePane = ({
     const DownloadIframe = 'downloadIframe';
 
     const [activeNode, setActiveNode] = useState(null);
+    const [selectedNodeForCopy, setSelectedNodeForCopy] = useState(null);
```
[the remark this answers is not in the document]
Comment (AbdelHedhili) [reply src/components/network-modification-tree-pane.js:84]: ok

File: src/components/graph/menus/create-node-menu.js [hunk 5]
```diff
@@ -55,6 +58,16 @@ const CreateNodeMenu = ({
         handleClose();
     }
 
+    function pasteNetworkModificationNode(insertMode) {
+        handlePasteNode(activeNode, insertMode);
```
[the remark this answers is not in the document]
Comment (AbdelHedhili) [reply src/components/graph/menus/create-node-menu.js:62]: ok

File: src/components/network-modification-tree-pane.js [hunk 6]
```diff
@@ -182,6 +184,27 @@ export const NetworkModificationTreePane = ({
...
     );
 
+    const handlePasteNode = useCallback(
+        (referenceNode, insertMode) => {
+            copyTreeNode(
+                studyUuid,
+                selectedNodeForCopy.id,
+                referenceNode.id,
+                insertMode
+            ).catch((errorMessage) => {
```
[the remark this answers is not in the document]
Comment (AbdelHedhili) [reply src/components/network-modification-tree-pane.js:194]: done

File: src/components/graph/menus/create-node-menu.js (hunk 1 above)
Comment (AbdelHedhili) [reply src/components/graph/menus/create-node-menu.js:125]: will do that at the end to make sure it doesn't break anything since it was here before my PR

Comment (AbdelHedhili) [reply src/components/graph/menus/create-node-menu.js]: fixed

**Review by etiennehomer — COMMENTED** [review]
We should be allowed to "paste node in a new branch" on root node nope ?
As we do for the insertion

File: src/components/graph/menus/create-node-menu.js [hunk 7]
```diff
@@ -90,6 +103,29 @@ const CreateNodeMenu = ({
...
+        COPY_MODIFICATION_NODE: {
+            onRoot: false,
+            action: () => copyNetworkModificationNode(),
+            id: 'copyNetworkModificationNode',
+        },
+        PASTE_MODIFICATION_NODE_ON_NEW_BRANCH: {
+            onRoot: false,
+            action: () => pasteNetworkModificationNode('NEW_BRANCH'),
+            id: 'pasteNetworkModificationNodeOnNewBranch',
+            disabled: selectedNodeForCopy == null,
```
Comment (etiennehomer) [inline src/components/graph/menus/create-node-menu.js:115]: === ?

File: src/components/graph/menus/create-node-menu.js [hunk 8]
```diff
@@ -90,6 +103,29 @@ const CreateNodeMenu = ({
...
+        },
+        PASTE_MODIFICATION_NODE_BEFORE: {
+            onRoot: false,
+            action: () => pasteNetworkModificationNode('BEFORE'),
+            id: 'pasteNetworkModificationNodeAbove',
+            disabled: selectedNodeForCopy == null,
```
Comment (etiennehomer) [inline src/components/graph/menus/create-node-menu.js:121]: same

File: src/components/graph/menus/create-node-menu.js [hunk 9]
```diff
@@ -90,6 +103,29 @@ const CreateNodeMenu = ({
...
+        },
+        PASTE_MODIFICATION_NODE_AFTER: {
+            onRoot: true,
+            action: () => pasteNetworkModificationNode('AFTER'),
+            id: 'pasteNetworkModificationNodeBelow',
+            disabled: selectedNodeForCopy == null,
```
Comment (etiennehomer) [inline src/components/graph/menus/create-node-menu.js:127]: same

**Review by etiennehomer — COMMENTED** [review]
Tests OK

File: src/components/graph/menus/create-node-menu.js (hunk 7 above)
Comment (AbdelHedhili) [reply src/components/graph/menus/create-node-menu.js:115]: nah i want to test for undefined and null

**Comment by AbdelHedhili** [comment]
> We should be allowed to "paste node in a new branch" on root node nope ? As we do for the insertion

actually we do not for insertion and so it's the same behavior here

**Comment by AbdelHedhili** [comment]
> We should be allowed to "paste node in a new branch" on root node nope ? As we do for the insertion

fixed

File: src/components/graph/menus/create-node-menu.js (hunk 7 above)
Comment (etiennehomer) [reply src/components/graph/menus/create-node-menu.js:115]: Ah yes

**Review by etiennehomer — APPROVED** [review]

**Review by FranckLecuyer — APPROVED** [review]

**Review by jonenst — APPROVED** [review]

**Pull request merged**